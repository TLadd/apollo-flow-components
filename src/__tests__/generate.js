jest.mock("fs");
jest.mock("glob");

const { generate } = require("../generate");
const fs = require("fs");
const glob = require("glob");

const aQueries = `
query GetList(
  $first: Int!
  $after: String
) {
  list(
    first: $first
    after: $after
  ) {
    pageInfo {
      totalCount
    }
  }
}

mutation DeleteItem($input: DeleteItemInput!) {
  deleteItem(input: $input) {
    success
  }
}        
`;

const bcQueries = `
fragment IdNameFrag on Record {
  id
  name
}

query GetRecord($id: ID!) {
  record(id: $id) {
    ...IdNameFrage
  }
} 

mutation UpdateRecord($input: UpdateRecordInput!) {
  updateRecord(input: $input) {
    record {
      ...IdNameFrage
    }
  }
} 
`;

const dQueries = `
query GetViewer {
  viewer {
    id
    name
  }
}      
`;

const eQueries = `
fragment IdNameFrag on Record {
  id
  name
}    
`;

const fQueries = `
mutation DeleteRecord($input: DeleteRecordInput!) {
  deleteRecord(input: $input) {
    success
  }
}
`;

describe("generate", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("generates expected output for flow", done => {
    glob.mockImplementationOnce((path, options, callback) => {
      callback(null, [
        "src/a/queries.graphql",
        "src/b/c/queries.graphql",
        "d/queries.graphql",
        "e/queries.graphql",
        "f/queries.graphql"
      ]);
    });

    fs.readFile.mockImplementation((path, options, callback) => {
      callback(
        null,
        {
          "src/a/queries.graphql": aQueries,
          "src/b/c/queries.graphql": bcQueries,
          "d/queries.graphql": dQueries,
          "e/queries.graphql": eQueries,
          "f/queries.graphql": fQueries
        }[path]
      );
    });

    fs.writeFile.mockImplementation((filename, fileText, callback) => {
      callback(null);
    });

    jest.spyOn(global.console, "warn").mockImplementation(() => {});

    generate({ target: "flow" }, error => {
      expect(fs.writeFile.mock.calls).toHaveLength(4);
      expect(global.console.warn).toHaveBeenCalledWith(
        "No queries or mutations found in e/queries.graphql"
      );

      const aCall = fs.writeFile.mock.calls.find(
        call => call[0] === "src/a/ApolloComps.js"
      );
      const bcCall = fs.writeFile.mock.calls.find(
        call => call[0] === "src/b/c/ApolloComps.js"
      );
      const dCall = fs.writeFile.mock.calls.find(
        call => call[0] === "d/ApolloComps.js"
      );
      const fCall = fs.writeFile.mock.calls.find(
        call => call[0] === "f/ApolloComps.js"
      );

      expect(error).toBeNull();
      expect(aCall[1]).toMatchSnapshot(
        "Generates a file with GetList query and DeleteItem mutation"
      );
      expect(bcCall[1]).toMatchSnapshot(
        "Generates a file with GetRecord query and UpdateRecord mutation"
      );
      expect(dCall[1]).toMatchSnapshot("Generates a file with GetViewer query");
      expect(fCall[1]).toMatchSnapshot(
        "Generates a file with DeleteRecord mutation"
      );

      done();
    });
  });

  it("generates expected output for typescript", done => {
    glob.mockImplementationOnce((path, options, callback) => {
      callback(null, [
        "src/a/queries.graphql",
        "src/b/c/queries.graphql",
        "d/queries.graphql",
        "e/queries.graphql",
        "f/queries.graphql"
      ]);
    });

    fs.readFile.mockImplementation((path, options, callback) => {
      callback(
        null,
        {
          "src/a/queries.graphql": aQueries,
          "src/b/c/queries.graphql": bcQueries,
          "d/queries.graphql": dQueries,
          "e/queries.graphql": eQueries,
          "f/queries.graphql": fQueries
        }[path]
      );
    });

    fs.writeFile.mockImplementation((filename, fileText, callback) => {
      callback(null);
    });

    jest.spyOn(global.console, "warn").mockImplementation(() => {});

    generate({ target: "typescript" }, error => {
      expect(fs.writeFile.mock.calls).toHaveLength(4);
      expect(global.console.warn).toHaveBeenCalledWith(
        "No queries or mutations found in e/queries.graphql"
      );

      const aCall = fs.writeFile.mock.calls.find(
        call => call[0] === "src/a/ApolloComps.tsx"
      );
      const bcCall = fs.writeFile.mock.calls.find(
        call => call[0] === "src/b/c/ApolloComps.tsx"
      );
      const dCall = fs.writeFile.mock.calls.find(
        call => call[0] === "d/ApolloComps.tsx"
      );
      const fCall = fs.writeFile.mock.calls.find(
        call => call[0] === "f/ApolloComps.tsx"
      );

      expect(error).toBeNull();
      expect(aCall[1]).toMatchSnapshot(
        "Generates a file with GetList query and DeleteItem mutation"
      );
      expect(bcCall[1]).toMatchSnapshot(
        "Generates a file with GetRecord query and UpdateRecord mutation"
      );
      expect(dCall[1]).toMatchSnapshot("Generates a file with GetViewer query");
      expect(fCall[1]).toMatchSnapshot(
        "Generates a file with DeleteRecord mutation"
      );

      done();
    });
  });

  it("calls callback with error if glob fails", done => {
    glob.mockImplementationOnce((path, options, callback) => {
      callback("Glob failed");
    });

    generate({ target: "typescript" }, error => {
      expect(error).toBe("Glob failed");
      done();
    });
  });

  it("calls callback with error if readFile fails", done => {
    glob.mockImplementationOnce((path, options, callback) => {
      callback(null, ["src/a/queries.graphql"]);
    });

    fs.readFile.mockImplementation((path, options, callback) => {
      callback("File does not exist");
    });

    generate({ target: "typescript" }, error => {
      expect(error).toBe("File does not exist");
      done();
    });
  });

  it("calls callback with error if writeFile fails", done => {
    glob.mockImplementationOnce((path, options, callback) => {
      callback(null, ["src/a/queries.graphql"]);
    });

    fs.readFile.mockImplementation((path, options, callback) => {
      callback(null, aQueries);
    });

    fs.writeFile.mockImplementation((filename, fileText, callback) => {
      callback("writeFile failed");
    });

    generate({ target: "flow" }, error => {
      expect(error).toBe("writeFile failed");
      done();
    });
  });
});
