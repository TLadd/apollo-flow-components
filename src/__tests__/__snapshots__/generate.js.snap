// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate generates expected output for flow: Generates a file with DeleteRecord mutation 1`] = `
"/* Generated using apollo-typed-components */
// @flow
import * as React from \\"react\\";
import { Mutation } from \\"react-apollo\\";
import {
  DeleteRecord
} from \\"./queries.graphql\\"
import type {
  DeleteRecord as DeleteRecordType,
  DeleteRecordVariables
} from \\"types\\";

class DeleteRecordMutationClass extends Mutation<DeleteRecordType, DeleteRecordVariables> {};

export const DeleteRecordMutation = (props: $Diff<React.ElementConfig<typeof DeleteRecordMutationClass>, { mutation: any }>) => <DeleteRecordMutationClass mutation={DeleteRecord} {...props} />;"
`;

exports[`generate generates expected output for flow: Generates a file with GetList query and DeleteItem mutation 1`] = `
"/* Generated using apollo-typed-components */
// @flow
import * as React from \\"react\\";
import { Query, Mutation, Subscription } from \\"react-apollo\\";
import {
  GetList,
  DeleteItem,
  OnDeleteItem
} from \\"./queries.graphql\\"
import type {
  GetList as GetListType,
  GetListVariables,
  DeleteItem as DeleteItemType,
  DeleteItemVariables,
  OnDeleteItem as OnDeleteItemType,
  OnDeleteItemVariables
} from \\"types\\";

class GetListQueryClass extends Query<GetListType, GetListVariables> {};
class DeleteItemMutationClass extends Mutation<DeleteItemType, DeleteItemVariables> {};
class OnDeleteItemSubscriptionClass extends Subscription<OnDeleteItemType, OnDeleteItemVariables> {};

export const GetListQuery = (props: $Diff<React.ElementConfig<typeof GetListQueryClass>, { query: any }>) => <GetListQueryClass query={GetList} {...props} />;
export const DeleteItemMutation = (props: $Diff<React.ElementConfig<typeof DeleteItemMutationClass>, { mutation: any }>) => <DeleteItemMutationClass mutation={DeleteItem} {...props} />;
export const OnDeleteItemSubscription = (props: $Diff<React.ElementConfig<typeof OnDeleteItemSubscriptionClass>, { subscription: any }>) => <OnDeleteItemSubscriptionClass subscription={OnDeleteItem} {...props} />;"
`;

exports[`generate generates expected output for flow: Generates a file with GetRecord query and UpdateRecord mutation 1`] = `
"/* Generated using apollo-typed-components */
// @flow
import * as React from \\"react\\";
import { Query, Mutation, Subscription } from \\"react-apollo\\";
import {
  GetRecord,
  UpdateRecord,
  OnUpdatedRecord
} from \\"./queries.graphql\\"
import type {
  GetRecord as GetRecordType,
  GetRecordVariables,
  UpdateRecord as UpdateRecordType,
  UpdateRecordVariables,
  OnUpdatedRecord as OnUpdatedRecordType,
  OnUpdatedRecordVariables
} from \\"types\\";

class GetRecordQueryClass extends Query<GetRecordType, GetRecordVariables> {};
class UpdateRecordMutationClass extends Mutation<UpdateRecordType, UpdateRecordVariables> {};
class OnUpdatedRecordSubscriptionClass extends Subscription<OnUpdatedRecordType, OnUpdatedRecordVariables> {};

export const GetRecordQuery = (props: $Diff<React.ElementConfig<typeof GetRecordQueryClass>, { query: any }>) => <GetRecordQueryClass query={GetRecord} {...props} />;
export const UpdateRecordMutation = (props: $Diff<React.ElementConfig<typeof UpdateRecordMutationClass>, { mutation: any }>) => <UpdateRecordMutationClass mutation={UpdateRecord} {...props} />;
export const OnUpdatedRecordSubscription = (props: $Diff<React.ElementConfig<typeof OnUpdatedRecordSubscriptionClass>, { subscription: any }>) => <OnUpdatedRecordSubscriptionClass subscription={OnUpdatedRecord} {...props} />;"
`;

exports[`generate generates expected output for flow: Generates a file with GetViewer query 1`] = `
"/* Generated using apollo-typed-components */
// @flow
import * as React from \\"react\\";
import { Query } from \\"react-apollo\\";
import {
  GetViewer
} from \\"./queries.graphql\\"
import type {
  GetViewer as GetViewerType
} from \\"types\\";

class GetViewerQueryClass extends Query<GetViewerType, {}> {};

export const GetViewerQuery = (props: $Diff<React.ElementConfig<typeof GetViewerQueryClass>, { query: any }>) => <GetViewerQueryClass query={GetViewer} {...props} />;"
`;

exports[`generate generates expected output for typescript: Generates a file with DeleteRecord mutation 1`] = `
"/* Generated using apollo-typed-components */
import * as React from \\"react\\";
import { Mutation } from \\"react-apollo\\";
import {
  DeleteRecord
} from \\"./queries.graphql\\"
import {
  DeleteRecord as DeleteRecordType,
  DeleteRecordVariables
} from \\"types\\";

type GetComponentProps<T> = T extends React.Component<infer P> ? P : never;
type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

class DeleteRecordMutationClass extends Mutation<DeleteRecordType, DeleteRecordVariables> {};

export const DeleteRecordMutation = (props: Omit<GetComponentProps<DeleteRecordMutationClass>, \\"mutation\\">) => <DeleteRecordMutationClass mutation={DeleteRecord} {...props} />;"
`;

exports[`generate generates expected output for typescript: Generates a file with GetList query and DeleteItem mutation 1`] = `
"/* Generated using apollo-typed-components */
import * as React from \\"react\\";
import { Query, Mutation, Subscription } from \\"react-apollo\\";
import {
  GetList,
  DeleteItem,
  OnDeleteItem
} from \\"./queries.graphql\\"
import {
  GetList as GetListType,
  GetListVariables,
  DeleteItem as DeleteItemType,
  DeleteItemVariables,
  OnDeleteItem as OnDeleteItemType,
  OnDeleteItemVariables
} from \\"types\\";

type GetComponentProps<T> = T extends React.Component<infer P> ? P : never;
type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

class GetListQueryClass extends Query<GetListType, GetListVariables> {};
class DeleteItemMutationClass extends Mutation<DeleteItemType, DeleteItemVariables> {};
class OnDeleteItemSubscriptionClass extends Subscription<OnDeleteItemType, OnDeleteItemVariables> {};

export const GetListQuery = (props: Omit<GetComponentProps<GetListQueryClass>, \\"query\\">) => <GetListQueryClass query={GetList} {...props} />;
export const DeleteItemMutation = (props: Omit<GetComponentProps<DeleteItemMutationClass>, \\"mutation\\">) => <DeleteItemMutationClass mutation={DeleteItem} {...props} />;
export const OnDeleteItemSubscription = (props: Omit<GetComponentProps<OnDeleteItemSubscriptionClass>, \\"subscription\\">) => <OnDeleteItemSubscriptionClass subscription={OnDeleteItem} {...props} />;"
`;

exports[`generate generates expected output for typescript: Generates a file with GetRecord query and UpdateRecord mutation 1`] = `
"/* Generated using apollo-typed-components */
import * as React from \\"react\\";
import { Query, Mutation, Subscription } from \\"react-apollo\\";
import {
  GetRecord,
  UpdateRecord,
  OnUpdatedRecord
} from \\"./queries.graphql\\"
import {
  GetRecord as GetRecordType,
  GetRecordVariables,
  UpdateRecord as UpdateRecordType,
  UpdateRecordVariables,
  OnUpdatedRecord as OnUpdatedRecordType,
  OnUpdatedRecordVariables
} from \\"types\\";

type GetComponentProps<T> = T extends React.Component<infer P> ? P : never;
type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

class GetRecordQueryClass extends Query<GetRecordType, GetRecordVariables> {};
class UpdateRecordMutationClass extends Mutation<UpdateRecordType, UpdateRecordVariables> {};
class OnUpdatedRecordSubscriptionClass extends Subscription<OnUpdatedRecordType, OnUpdatedRecordVariables> {};

export const GetRecordQuery = (props: Omit<GetComponentProps<GetRecordQueryClass>, \\"query\\">) => <GetRecordQueryClass query={GetRecord} {...props} />;
export const UpdateRecordMutation = (props: Omit<GetComponentProps<UpdateRecordMutationClass>, \\"mutation\\">) => <UpdateRecordMutationClass mutation={UpdateRecord} {...props} />;
export const OnUpdatedRecordSubscription = (props: Omit<GetComponentProps<OnUpdatedRecordSubscriptionClass>, \\"subscription\\">) => <OnUpdatedRecordSubscriptionClass subscription={OnUpdatedRecord} {...props} />;"
`;

exports[`generate generates expected output for typescript: Generates a file with GetViewer query 1`] = `
"/* Generated using apollo-typed-components */
import * as React from \\"react\\";
import { Query } from \\"react-apollo\\";
import {
  GetViewer
} from \\"./queries.graphql\\"
import {
  GetViewer as GetViewerType
} from \\"types\\";

type GetComponentProps<T> = T extends React.Component<infer P> ? P : never;
type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

class GetViewerQueryClass extends Query<GetViewerType, {}> {};

export const GetViewerQuery = (props: Omit<GetComponentProps<GetViewerQueryClass>, \\"query\\">) => <GetViewerQueryClass query={GetViewer} {...props} />;"
`;
